clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -1392,4 +1402,16 @@ public void registerStateListener(MemberStateListener listener) {
         }
         this.stateUpdatesListeners.add(listener);
     }
+
+    /**
+     * If either a new target assignment or new metadata is available that we have not yet attempted
+     * to reconcile, and we are currently in state RECONCILING, trigger reconciliation.
+     */
+    @Override
+    public PollResult poll(final long currentTimeMs) {
+        if (state == MemberState.RECONCILING && attemptReconciliation) {
+            reconcile();
```

Comment:
I think doing this in `poll` isn't strictly necessary. We could call `reconcile` in 3 places (on metadata update, on new assignment, after an ongoing reconciliation finished), and wouldn't have to rely on `poll` to attempt the reconciliation. That would  be mostly the same change, but simpler. But since the ticket was written that way, for now I implemented it inside poll.

Reply:
For the context, I think that the idea was to move away from the edge reconciliation triggers to a model where we would check if the current assignment matches the target assignment in `poll` and reconcile if it does not and there is no on-going reconciliation. The benefits of this is that it works in all cases even if we, for instance, miss a trigger.

Reply:
Got it. But if we completely move away from triggers (so remove `attemptReconciliation` boolean in this change) and re-attempt reconciliation in every poll, we'd stay in a reconciliation loop as long as topic metadata is missing. Do we know that missing target IDs are always resolved in a relatively short amount of time? Otherwise, we'd need a back-off mechanism.

Reply:
Regarding the reconcile on poll, agree with @dajac that we were moving away from the multiple triggers looking for a more unified approach, that would reconcile not based on different events, but rather on a regular check of the target assignment, that could be affected by all those events. So with this, we have a `reconcile()` as the single place to determine if we should reconcile ([this initial block](https://github.com/apache/kafka/blob/2158ab7b66e0b57bdaf7872511e48ff076160320/clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java#L787-L822)),  
1. no reconciliation in progress 
2. target assignment different from the current assignment
3. resolved target different from the current assignment

In that sense, calling that reconcile regularly on every poll is probably all we want, and aligns with what @dajac described, as it will run the checks above and proceed. But the `attemptReconciliation` introduces something extra that maybe we can simplify a bit.

If I get it right, all we want with that attempt logic is to avoid going through the unresolved assignment and requesting metadata if we know we were waiting for it already and it hasn't been received, right? In that sense, it looks to me as an internal improvement to how we resolve metadata in the `[findResolvableAssignmentAndTriggerMetadataUpdate](https://github.com/apache/kafka/blob/2158ab7b66e0b57bdaf7872511e48ff076160320/clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java#L1012)`, and not as another condition to reconcile. The `findResolvableAssignmentAndTriggerMetadataUpdate` could consider if we are awaiting metadata for a target assignment before checking the target and requesting it again. It would be no-op if we are awaiting metadata for the target, and the same common reconcile checks (check # 3) would ensure that a reconciliation is not triggered if resolved target didn't change.

Reply:
Good points. Let's keep it simple for now and remove `attemptReconciliation` and the metadata update listener, and just reconcile on every poll until the metadata is complete. We can always bring back the extra complexity if this causes problems.